### ISSUE TITLE
Add support to correlate analysis server notifications with document versions

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#30301]
Currently when you get a notification from the analysis server, it's hard to know what version of your document it refers to. When a user is typing and causing lots of buffer changes, the notification may be "stale" by the time it's received. This means that all offsets could be incorrect. If using this for rendering something real-time in the editor (such as syntax highlighting or other editor decorations) this might look really quirky.

I haven't had any issues as a result of this yet in Dart Code, though suspect it may affect a feature like [these decorations](https://github.com/Dart-Code/Dart-Code/issues/383). I don't know if this may already be a solved issue by other editors, but I can't think of how to reliably resolve it without having some correlation between the `updateContent` calls and the notifications (such as a document version either transmitted or returned in its response, and then also included in a notification that relates to that file).

(cc @bwilkerson)

**Comment by bwilkerson** [comment]
We used to return a modification stamp in an attempt to address this issue, but that was recently broken and no clients reported a problem, so we assumed that clients were not using it and haven't fixed it.

Even with that there is still a possibility of the two getting out of sync. That's the reason that the server API uses the concept of eventual consistency: if there is enough time between changes to a file the data returned by server will eventually be consistent with the client's view of the file.

> I haven't had any issues as a result of this yet in Dart Code ...

Given that it hasn't caused problems yet I'm tempted to postpone fixing it until we know it's a problem, but adding a client-managed id passed in with `updateContent` is a promising direction. It won't handle the case of changes being made outside the client, but that might be less frequent and hence less interesting.

**Comment by DanTup** [comment]
> We used to return a modification stamp in an attempt to address this issue, but that was recently broken and no clients reported a problem, so we assumed that clients were not using it and haven't fixed it.

Ah, I remember the discussion - it didn't occur to me why it'd be useful at the time! =)

> Given that it hasn't caused problems yet I'm tempted to postpone fixing it until we know it's a problem

In my decorations prototype it actually caused a lot of jumping around so I had to put a timeout in to delay the frequency of updates. I'm happy to get a proper working implementation of that first though; it'll be a good example to review and test out changes with.

**Comment by bwilkerson** [comment]
Note that the performance characteristics are likely to change significantly when this functionality is moved to server, so it might not be worth spending time tuning the prototype.

**Comment by DanTup** [comment]
Yeah sorry, I meant we should get it moved to the server and then test that, not test with the current one. The current implementation is certainly failing badly here since it's sending a getHover for every decoration - this means lots of round trips, increasing the chances of stale data.

**Comment by DanTup** [comment]
@bwilkerson After using an analyzer notification the flicker is still the same. It's not so much the performance as just the fact that you can get a keypress in between the `updateContent` request being sent and the notification coming back which means everything is a character out. You see the decorations "dancing" by one character.

![Flicker](https://user-images.githubusercontent.com/1078012/28893443-52fc93ac-77c9-11e7-8529-1754971fb552.gif)

It can be partly mitigated by the client by doing what I did with the prototype, but some correlation/version would make this solid. If you have any other ideas, I'm open to giving them a shot :-)

**Comment by bwilkerson** [comment]
I agree that this is less than ideal. I'm open to the idea of adding some kind of time-stamp data, but it might not be trivial to either design or implement, and isn't currently on the schedule for this quarter. @devoncarew for visibility of the possible additional work.

> ... I had to put a timeout in to delay the frequency of updates ...

I'm not sure what kind of timeout you added so this might be the same as what you've already done, but one possible short-term work-around would be to wait some small amount of time after the last key press before applying the most recently received notification results. That would basically give the server time to respond with up-to-date data before you update the display.

Another possibility, given that VSCode moves the decorations with the text, would be to compare the previously received decorations with the new decorations. You can choose to not display the new decorations if there are the same number of decorations, all the decorations before the changed code are the same, and all the decorations after the change are the same except for being a fixed offset from where the old decorations were.

**Comment by DanTup** [comment]
@bwilkerson That's kinda what I did.. On every notification I reset the timer for 500ms and only perform the update if there weren't any new notifications for 500ms. This should work fine if every notification comes less than 500ms than the keypress that triggered it, though I don't know how likely that is for large projects. For this data, response time isn't that critical, so this value could be increased if it's an issue (as noted, Code moves the decorations correctly on buffer edits so the negatives are not huge).

Most of the other notifications include offsets that don't result in anything rendering so this isn't really an issue; though I suspect if anyone is using Highlights for syntax highlighting they'd probably see this issue (I currently don't, though did plan to switch to it when Code supports it if it handles string interpolation better than the regex syntax we have).

I think the workaround with the timer is reasonable (and if not, your other idea seems reasonable), so I don't think this is that important right now, but it's worth having on the list; it might come up again if other editors implement these decorations!

**Comment by bwilkerson** [comment]
> ... if anyone is using Highlights for syntax highlighting ...

I believe that IntelliJ is using highlights.

> ... if it handles string interpolation better than the regex syntax we have ...

It handles string interpolation correctly. Whether that's better than what you have is another question. It also provides semantic highlighting, so static members and instance members, for example, could be highlighted differently if you want.

> ... it's worth having on the list ...

Absolutely. Thanks!

**Comment by devoncarew** [comment]
@DanTup, I wonder how quickly Dart-Code is sending overlay updates? From that gif, it looks like it's being sent appropriately per keystroke. For other clients I've written, I have done something to batch up changes. So, debounce changes by 200ms, and flush after a while even if we still are receiving changes. That would wait to send an overlay until the user pauses in their typing, and would make the above overlays/decorations a bit less twitchy.

You still want to flush any unsent changes if you make an explicit request, like for code completions.

Batching changes up slightly would also have benefits in terms of CPU usage - you're asking the analysis server to compute 1/10 the number of changes.

**Comment by DanTup** [comment]
Interesting - I had a search through their code on GH but can't find anything using the notification to see whether there's any similar workaround - maybe @devoncarew knows whether it has this issue and/or another workaround?

> It handles string interpolation correctly. Whether that's better than what you have is another question. It also provides semantic highlighting, so static members and instance members, for example, could be highlighted differently if you want.

It definitely sounds better - our current one is even using casing of the first letter to guess whether things are class names or functions (which really bugs me). I think we borrowed it from Atom =D

**Comment by devoncarew** [comment]
I've not worked w/ the semantic highlighting code myself, but the code would be here: https://github.com/JetBrains/intellij-plugins/tree/master/Dart.

**Comment by DanTup** [comment]
> For other clients I've written, I have done something to batch up changes. So, debounce changes by 200ms, and flush after a while even if we still are receiving changes.

@devoncarew I can see logic in that, though I think it could also make some things feel more sluggish. I think it also only reduces the issue by doing it less rather than actually improving it. I think delaying just the decorations updating would be better since it wouldn't have a negative impact on other things.

Ofc, if anyone reports high CPU use of the analyzer, I would reconsider (though I'd also raise the question of whether the analyzer should support doing this internally; it feels like performance optimisations for its own work should be hidden from the client; otherwise the client needs to have an understanding of its internal performance characteristics).

> I've not worked w/ the semantic highlighting code myself, but the code would be here: https://github.com/JetBrains/intellij-plugins/tree/master/Dart.

That's where I was looking but couldn't find much. Just found [this](https://github.com/JetBrains/intellij-plugins/blob/3fd143aecedaa817f36e83727892000d54365ad8/Dart/src/com/jetbrains/lang/dart/analyzer/DartServerData.java#L80) now which contains this:

> if (myFilePathsWithUnsentChanges.contains(filePath)) return;

This seems more likely a fix for what you mentioned above (buffering up changes) than the original issue I raised. But I did find that it calls into [this](https://github.com/JetBrains/intellij-plugins/blob/581d3939d0a6017155023f274b539f268b909040/Dart/src/com/jetbrains/lang/dart/analyzer/FileOffsetsManager.java#L74) which seems to be doing some mapping of offsets and:

> if (offsets != null && file.getModificationStamp() == offsets.myFileModificationStamp) {
>    return offsets;
> }

Though I couldn't figure out if that was related to the analyzer timestamp (the one that got broken) or something else. Maybe I'll dig into it more if Code does get dynamic colouring support and I end up seeing this same issue to see how they mitigate it.

**Comment by DanTup** [comment]
FWIW, this request is probably not really relevant for VS Code anymore. In LSP, most things are pulled from the client rather than pushed, and whenever a document is modified, VS Code (or any other LSP client) should cancel and discard results for any requests sent prior to the modification.

While we do still use notifications for Closing Labels today, in the LSP server we are able to capture the LSP document version at the start of a request to include in notifications, although the example above (Closing Labels) will soon migrate to LSP's new Inlay Hints so that's not worthwhile.

@bwilkerson unless there's some value in this for IntelliJ or other API consumers, this could be closed.

**Comment by bwilkerson** [comment]
I'm not aware of any issues in IntelliJ, so I'm going to close it.
